packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/PrimitiveTest.m

```diff
@@ -0,0 +1,29 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <Flutter/Flutter.h>
+#import <XCTest/XCTest.h>
+#import "EchoMessenger.h"
+#import "primitive.gen.h"
+
+///////////////////////////////////////////////////////////////////////////////////////////
+@interface PrimitiveTest : XCTestCase
+@end
+
+///////////////////////////////////////////////////////////////////////////////////////////
+@implementation PrimitiveTest
+
+- (void)testPrimitive {
+  EchoBinaryMessenger* binaryMessenger = [[EchoBinaryMessenger alloc] init];
+  PrimitiveFlutterApi* api = [[PrimitiveFlutterApi alloc] initWithBinaryMessenger:binaryMessenger];
+  XCTestExpectation* expectation = [self expectationWithDescription:@"callback"];
+  [api inc:@1
+      completion:^(NSNumber* _Nonnull result, NSError* _Nullable err) {
+        XCTAssertEqualObjects(@1, result);
+        [expectation fulfill];
+      }];
+  [self waitForExpectations:@[ expectation ] timeout:1.0];
+}
+
```

Comment:
Why aren't the other primitives tested, as they are on the Java side?

Reply:
Test Driven Development, I just wrote the tests that would lead to the code changes I wanted.  I knew I didn't need them for objc because I was tapping into a function that was already written and tested.  On the Java side I was calling into a newly written `_boxedType` function which didn't have a full implementation (because it was a duplication of `_javaTypeForDartType` logic which was a mistake).

Reply:
> Test Driven Development, I just wrote the tests that would lead to the code changes I wanted.

This is the opposite of every description of TDD I've encountered. The standard TDD practice is that you write test cases for all of your requirements (which in this case is all of the primitive types working) and then write as little *implementation code* as possible to make those tests pass (which in this case happens to be trivial). Minimizing the amount of testing definitely isn't a TDD goal.

Starting with knowledge of the implementation internals to decide what tests you should write isn't actually test-driven, and the result is that your tests aren't actually asserting everything you want to be true. If someone later refactored in a way that it wasn't sharing that function any more, and broke all the other primitive types, no tests would fail, because you've never asserted that they work.

If Pigeon is supporting all the primitive types in all languages, then it should have tests covering all primitive types in all languages.

Reply:
There is already another test for all the datatypes, it's called all_datatypes.  This was duplicating those tests and forging a duplicate path through the code with `_boxedType`, which is removed in a later PR.

The ideal for google's code is 100% test coverage.  By writing the test to assert the code before writing the code you are getting a higher test coverage.

Reply:
> The ideal for google's code is 100% test coverage.

If you mean LoC, that's a necessary (since any code that's not executed by tests isn't getting any coverage, obviously) but not sufficient condition for good testing.

This code has 100% test coverage by lines:

```
test_absolute_value() {
  assert(absolute_value(-1), 1);
}
```

```
int absolute_value(int value) {
  return value.abs();
}
```

I'm saying that if you start from the assumption that you are going to implement `absolute_value` by calling `abs`, and use that to justify only writing the one test above, that's not good test design, because:

```
int absolute_value(int value) {
  return value * -1;
}
```

still passes all tests (and even has 100% coverage!).

> By writing the test to assert the code before writing the code you are getting a higher test coverage.

I'm not sure what the relevance of this statement is to my comments. I did not suggest that one shouldn't write the tests before the code, I said one should write tests that assert all of the requirements, rather than simply writing the test to provide minimal coverage based on knowledge of the implementation.

> There is already another test for all the datatypes

A test for this specific code, or a test for code at another layer that this code is implemented in terms of? If the latter, see the example above; the existence of tests for `abs` doesn't guarantee correctness of `absolute_value`. If the former, then why does this new test exist at all?

Reply:
Yea, I see what you are saying.  It's an interesting discussion and I see your point.  My goal is 100% test coverage, `value * -1` is fine if we only send in negative numbers.  The other side of the coin is building the tests like you are suggesting is making changes to the code more difficult, increasing engineering cost and maintenance cost.  Pigeon isn't in its lifecycle where that makes sense yet, to me.  We can chat about his on the phone later.

Reply:
> `value * -1` is fine if we only send in negative numbers.

The title of this PR is "implement primitive datatypes" though, not "implement support for 'int'".

Reply:
Talked offline, it makes sense to add the tests like Android has.  FWIW they were getting tested that they compile, not tested that they execute.

Reply:
Done, added objc tests.